consul/resource_service_defaults_config_entry.go

```diff
@@ -0,0 +1,557 @@
+			formattedSetValue, err := formatKeys(valueList[0], formatKey)
+			if err != nil {
+				return nil, err
+			}
+			return formattedSetValue, nil
+		}
+		return nil, nil
+	}
+	return config, nil
+}
+
+func resourceConsulServiceDefaultsConfigEntryUpdate(d *schema.ResourceData, meta interface{}) error {
+	client, qOpts, wOpts := getClient(d, meta)
+	configEntries := client.ConfigEntries()
+
+	name := d.Get("name").(string)
+
+	configMap := make(map[string]interface{})
+	configMap["kind"] = "service-defaults"
+
+	configMap["name"] = name
+
+	kind := configMap["kind"].(string)
+	err := d.Set("kind", kind)
+
+	if err != nil {
+		return err
+	}
+
+	fixQOptsForServiceDefaultsConfigEntry(name, kind, qOpts)
+
+	var attributes []string
+
+	for key, _ := range serviceDefaultsConfigEntrySchema {
+		attributes = append(attributes, key)
+	}
+
+	for _, attribute := range attributes {
+		configMap[attribute] = d.Get(attribute)
+	}
+
+	formattedMap, err := formatKeys(configMap, formatKey)
+	if err != nil {
+		return err
+	}
+
+	configEntry, err := makeServiceDefaultsConfigEntry(kind, name, formattedMap.(map[string]interface{}), wOpts.Namespace, wOpts.Partition)
+	if err != nil {
+		return err
+	}
+
+	if _, _, err := configEntries.Set(configEntry, wOpts); err != nil {
+		return fmt.Errorf("failed to set '%s' config entry: %v", name, err)
+	}
+	_, _, err = configEntries.Get(configEntry.GetKind(), configEntry.GetName(), qOpts)
+	if err != nil {
+		if strings.Contains(err.Error(), "Unexpected response code: 404") {
+			return fmt.Errorf(`failed to read config entry after setting it.
+This may happen when some attributes have an unexpected value.
+Read the documentation at https://www.consul.io/docs/agent/config-entries/%s.html
+to see what values are expected`, configEntry.GetKind())
+		}
+		return fmt.Errorf("failed to read config entry: %v", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s-%s", kind, name))
+	return resourceConsulServiceDefaultsConfigEntryRead(d, meta)
+}
+
+func resourceConsulServiceDefaultsConfigEntryRead(d *schema.ResourceData, meta interface{}) error {
+	client, qOpts, _ := getClient(d, meta)
+	configEntries := client.ConfigEntries()
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	fixQOptsForConfigEntry(configName, configKind, qOpts)
+
+	_, _, err := configEntries.Get(configKind, configName, qOpts)
+	return err
+}
+
+func resourceConsulServiceDefaultsConfigEntryDelete(d *schema.ResourceData, meta interface{}) error {
+	client, _, wOpts := getClient(d, meta)
+	configEntries := client.ConfigEntries()
+	configKind := d.Get("kind").(string)
+	configName := d.Get("name").(string)
+
+	if _, err := configEntries.Delete(configKind, configName, wOpts); err != nil {
+		return fmt.Errorf("failed to delete '%s' config entry: %v", configName, err)
+	}
+	d.SetId("")
+	return nil
+}
+
+func makeServiceDefaultsConfigEntry(kind, name string, configMap map[string]interface{}, namespace, partition string) (consulapi.ConfigEntry, error) {
```

Comment:
I don't think `kind` needs to be passed as a param to this method. You could move the string `service-defaults` as a constant and reference it whenever needed

Reply:
ok